packages/trie/src/util/asyncWalk.ts

```diff
@@ -0,0 +1,60 @@
+import { RLP } from '@ethereumjs/rlp'
+import { equalsBytes, toHex } from 'ethereum-cryptography/utils'
+
+import { BranchNode } from '../node/branch.js'
+import { ExtensionNode } from '../node/extension.js'
+
+import type { Trie } from '../trie'
+import type { TrieNode } from '../types'
+
+export type NodeFilter = (node: TrieNode, key: number[]) => Promise<boolean>
+export type OnFound = (node: TrieNode, key: number[]) => Promise<any>
+
+/**
+ * Walk Trie via async generator
+ * @param nodeHash - The root key to walk on.
+ * @param currentKey - The current (partial) key.
+ * @param onFound - Called on every node found (before filter)
+ * @param filter - Filter nodes yielded by the generator.
+ * @param visited - Set of visited nodes
+ * @returns AsyncIterable<{ node: TrieNode; currentKey: number[] }>
+ * Iterate through nodes with
+ * `for await (const { node, currentKey } of trie._walkTrie(root)) { ... }`
+ */
+export async function* _walkTrie(
+  this: Trie,
+  nodeHash: Uint8Array,
+  currentKey: number[] = [],
+  onFound: OnFound = async (_trieNode: TrieNode, _key: number[]) => {},
+  filter: NodeFilter = async (_trieNode: TrieNode, _key: number[]) => true,
+  visited: Set<string> = new Set<string>()
+): AsyncIterable<{ node: TrieNode; currentKey: number[] }> {
+  if (equalsBytes(nodeHash, this.EMPTY_TRIE_ROOT)) {
+    return
+  }
+  try {
+    const node = await this.lookupNode(nodeHash)
+    if (node === undefined || visited.has(toHex(this.hash(node!.serialize())))) {
+      return
+    }
+    visited.add(toHex(this.hash(node!.serialize())))
+    await onFound(node!, currentKey)
+    if (await filter(node!, currentKey)) {
+      yield { node: node!, currentKey }
+    }
+    if (node instanceof BranchNode) {
+      for (const [nibble, childNode] of node._branches.entries()) {
+        const nextKey = [...currentKey, nibble]
+        const _childNode: Uint8Array =
+          childNode instanceof Uint8Array ? childNode : this.hash(RLP.encode(childNode))
+        yield* _walkTrie.bind(this)(_childNode, nextKey, onFound, filter, visited)
+      }
+    } else if (node instanceof ExtensionNode) {
+      const childNode = node.value()
+      const nextKey = [...currentKey, ...node._nibbles]
+      yield* _walkTrie.bind(this)(childNode, nextKey, onFound, filter, visited)
+    }
+  } catch (e) {
```

Comment:
Why apply this try catch on the entire function and not log the error. Don't want to evolve myself into coding standards of ethereumjs too much but IMO a good practice is to do 1 statement per try catch and then match the statement's respective errors or allow errors to bubble up further.

Reply:
Really the only potential error here is something like 
"DB: Node Missing".  We want the walk to ignore these errors and keep going, and I would prefer it happened silently.

I don't want the error to throw, because that stops the walk from happening.  I want it to simply continue on.

The reason here is that we want to walk through a sparse trie -- a Trie that has been built from a proof.  
A walk through a trie like this will find a hash on the branch of a proof node for a node that is not in the database, 

said node was not part of the proof, but it's hash was part of the RLP encoding of one of the proof nodes.  So the walk will look for that hash in the DB, and the DB will throw an error.  Which in turn would stop the walk prematurely, even though it is a valid Trie with nodes remaining.